bin/generated_template.d.ts

```diff
@@ -0,0 +1,113 @@
+/** Provides information pointing to a location within a source. */
+export interface Location {
+  /** Line in the parsed source (1-based). */
+  line: number;
+  /** Column in the parsed source (1-based). */
+  column: number;
+  /** Offset in the parsed source (0-based). */
+  offset: number;
+}
+
+/** The `start` and `end` position's of an object within the source. */
+export interface LocationRange {
+  /** Any object that was supplied to the `parse()` call as the `grammarSource` option. */
+  source: any;
+  /** Position at the beginning of the expression. */
+  start: Location;
+  /** Position after the end of the expression. */
+  end: Location;
+}
+
+export interface LiteralExpectation {
+  type: "literal";
+  text: string;
+  ignoreCase: boolean;
+}
+export interface ClassParts extends Array<string | ClassParts> {
+}
+export interface ClassExpectation {
+  type: "class";
+  parts: ClassParts;
+  inverted: boolean;
+  ignoreCase: boolean;
+}
+export interface AnyExpectation {
+  type: "any";
+}
+export interface EndExpectation {
+  type: "end";
+}
+export interface OtherExpectation {
+  type: "other";
+  description: string;
+}
+export type Expectation =
+  | AnyExpectation
+  | ClassExpectation
+  | EndExpectation
+  | LiteralExpectation
+  | OtherExpectation;
+
+declare class _PeggySyntaxError extends Error {
+  /**
+   * Constructs the human-readable message from the machine representation.
+   *
+   * @param expected Array of expected items, generated by the parser
+   * @param found Any text that will appear as found in the input instead of
+   *   expected
+   */
+  static buildMessage(expected: Expectation[], found: string | null): string;
+  message: string;
+  expected: Expectation[];
+  found: string | null;
+  location: LocationRange;
+  name: string;
+  constructor(
+    message: string,
+    expected: Expectation[],
+    found: string | null,
+    location: LocationRange,
+  );
+  format(sources: {
+    source?: any;
+    text: string;
+  }[]): string;
+}
+
+export interface ParserTracer {
+  trace(event: ParserTracerEvent): void;
+}
+
+export type ParserTracerEvent
+  = { type: "rule.enter"; rule: string; location: LocationRange }
+  | { type: "rule.fail"; rule: string; location: LocationRange }
+  | { type: "rule.match"; rule: string; result: any; location: LocationRange };
+
+export type StartRules = $$$StartRules$$$;
+export interface ParseOptions<T extends  StartRules = $$$DefaultStartRule$$$> {
+  /**
+   * Object that will be attached to the each `LocationRange` object created by
+   * the parser. For example, this can be path to the parsed file or even the
+   * File object.
+   */
+  grammarSource?: any;
+  startRule?: T;
+  tracer?: ParserTracer;
+
+  // Internal use only:
+  peg$library?: boolean;
+  // Internal use only:
+  peg$currPos?: number;
+  // Internal use only:
+  peg$silentFails?: number;
+  // Internal use only:
+  peg$maxFailExpected?: Expectation[];
```

Comment:
Pedantic; nearly no library I've seen actually accounts for compiler flags like these.

---

_Technically_ there is one thing here; there's a compiler flag called [`exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes). At runtime there is actually a distinction between `{}` and something like `{ tracer: undefined }` - that being the result of `Object.keys`, `in`, etc.

When it's enabled there's therefore then a distinction between `{ x?: number }` and `{ x?: number | undefined }`. So to be more convenient for users who use `exactOptionalPropertyTypes` you could write an explicit `| undefined` on all of your optional _properties._

Note: this does NOT apply the function _parameters_ only to object _properties._ In other no need to ever say `foo(x?: number | undefined)` since the shorter `foo(x?: number)` is equivalent.

Reply:
Nod.

Reply:
Doesn't this mean you can pass in `undefined` rather than just omitting the field?  I don't want that for the internal-use-only fields.

Reply:
I can't mark this as resolved but it can be.